### PR TITLE
Upgrade to Node 18.

### PR DESCRIPTION
**Pull request by ray-lee**
## Description

Upgrades this project to build with Node 18, along with some other quality/security improvements. This includes:

- Upgrading dependencies that were incompatible with Node 18
- Upgrading dependencies to resolve `npm audit` security errors (not all of them, just the easy ones for now)
- Code/configuration changes required by major version upgrades of dependencies, including
  - Major typescript update required adding some missing type parameters to Promise constructors
  - Major webpack update required updating some webpack config files, moving some npm script arguments into webpack config files, and removing some no-longer-supported arguments
- Some minor formatting changes to resolve lint errors
- Additional `export`s from some files to resolve documentation generation warnings

## Motivation and Context

The project had been using Node 12, which is no longer supported, and will not be allowed in GitHub actions soon. Notion: https://www.notion.so/lyrasis/Upgrade-web-opds-client-to-Node-18-44f9a03cbb0545cc863f19a5db8bb9c0

## How Has This Been Tested?

- Verify that `npm install` runs without error
- Verify that all npm scripts in all package.json files (package.json, packages/server/package.json, and packages/web-opds-client/package.json) run without error (except for `npm publish`, which we'll test when we actually publish). This especially includes `npm test` in packages/web-opds-client.
- Verify that the app runs in dev:
  - `npm run dev`
  - Open http://localhost:3000 in a browser
  - Enter a feed URL, e.g. https://cerberus.tpp-qa.lyrasistechnology.org/cerberus-test-1/feed/1206
  - Verify the feed is displayed

## Checklist:

- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

### CODE REVIEW

File: packages/web-opds-client/tsconfig.json [hunk 1]
```diff
@@ -14,6 +14,7 @@
   },
   "exclude": ["node_modules", "lib", "dist"],
   "typedocOptions": {
+    "entryPointStrategy": "expand",
```
Comment (tdilauro) [inline packages/web-opds-client/tsconfig.json:17]: Explanation: "expand" behavior was the default before v0.22. For v0.22+, "resolve" behavior is the default. We're updating from v0.20... to v0.23..., so this change is to retain the "expand" behavior.

From [the docs](https://typedoc.org/api/enums/EntryPointStrategy.html):
> The default behavior in v0.21 and earlier. Behaves like the resolve behavior, but will recursively expand directories into an entry point for each file within the directory.

**Review by tdilauro — COMMENTED** [review]
This looks good and seems to work fine.

I do have one question about the husky dependency being reverted to an older version. Could you clarify?

File: package.json [hunk 2]
```diff
@@ -21,9 +25,9 @@
     "build-docs": "cd packages/web-opds-client && npm run build-docs"
   },
   "devDependencies": {
-    "husky": "^6.0.0",
-    "lint-staged": "^9.4.3",
-    "prettier": "1.19.1"
+    "husky": "^4.3.0",
```
[the remark this answers is not in the document]
Comment (ray-lee) [reply package.json:28]: Yes, version 6 wasn't actually working -- there is a pre-commit hook that is supposed to run, but it doesn't -- so I backed it off to the same version that is used in circulation-admin. That got the hook working. We can upgrade it in all the projects in a separate ticket, but there isn't any pressing reason to do so at the moment.

**Comment by ray-lee** [comment]
@tdilauro Could you approve this? I think everything has been resolved.

**Review by tdilauro — APPROVED** [review]
Looks good! 🚀

**Pull request merged**